[SPARK-11763] [ML] Add save,load to LogisticRegression Estimator

jkbradley opened the pull request:
Add save/load to LogisticRegression Estimator, and refactor tests a little to make it easier to add similar support to other Estimator, Model pairs.

Moved LogisticRegressionReader/Writer to within LogisticRegressionModel

CC: @mengxr

jkbradley committed:
Added save/load to LogisticRegression Estimator, and improved unit test for it

jkbradley committed:
Moved LogisticRegressionReader/Writer to within LogisticRegressionModel

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/util/DefaultReadWriteTest.scala:
```diff
@@ -64,6 +66,47 @@ trait DefaultReadWriteTest extends TempDirectory { self: Suite =>
+    val estimator2 = testDefaultReadWrite(estimator)
+    testParams.foreach { case (p, v) =>
+      val param = estimator.getParam(p)
+      assert(estimator.get(param).get === estimator2.get(param).get)
```
This is awkward, but type issues prevented me from using ```apply``` instead of ```get```

jkbradley commented on mllib/src/test/scala/org/apache/spark/ml/util/DefaultReadWriteTest.scala:
```diff
@@ -64,6 +66,47 @@ trait DefaultReadWriteTest extends TempDirectory { self: Suite =>
+      assert(estimator.get(param).get === estimator2.get(param).get)
+    }
+
+    deleteTempDir()
```
Question: Am I right that this is blocking and not risking flakiness?  I'm guessing based on [https://github.com/apache/spark/blob/fd14936be7beff543dbbcf270f2f9749f7a803c4/core/src/main/scala/org/apache/spark/util/Utils.scala#L889] being synchronized.

mengxr replied:
Shall we modify `testDefaultReadWrite` to append some random string to UID so calls to it are independent?

SparkQA commented:
**[Test build #46032 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46032/consoleFull)** for PR 9749 at commit [`528b5b1`](https://github.com/apache/spark/commit/528b5b127655e9d9212fe985376b794828a292da).

jkbradley commented:
Oops, I just realized I forgot to save the parent field in both this and Pipeline.  Shall we just document that it does not get saved, and handle it for 1.7?

SparkQA commented:
**[Test build #46032 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46032/consoleFull)** for PR 9749 at commit [`528b5b1`](https://github.com/apache/spark/commit/528b5b127655e9d9212fe985376b794828a292da).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:\n  * `class BitSet(numBits: Int) extends Serializable `\n

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46032/
Test PASSed.

mengxr commented on mllib/src/test/scala/org/apache/spark/ml/util/TempDirectory.scala:
```diff
@@ -39,7 +39,12 @@ trait TempDirectory extends BeforeAndAfterAll { self: Suite =>
   }
+
+  /** Delete [[tempDir]] */
+  def deleteTempDir(): Unit = {
```
We won't need this if we implement testDefaultReadWrite properly.

mengxr commented:
LGTM except one minor issue commented inline.

jkbradley committed:
Changed testDefaultReadWrite to create a subdir in tempDir with a random name, to avoid conflicts between calls

jkbradley commented:
Thanks, updated!

SparkQA commented:
**[Test build #46102 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46102/consoleFull)** for PR 9749 at commit [`54a4885`](https://github.com/apache/spark/commit/54a4885f1626046bf7fcf4da66abf3ae47314f32).

SparkQA commented:
**[Test build #46102 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46102/consoleFull)** for PR 9749 at commit [`54a4885`](https://github.com/apache/spark/commit/54a4885f1626046bf7fcf4da66abf3ae47314f32).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/46102/
Test FAILed.

jkbradley committed:
removed unnecessary part of PipelineSuite read/write test

jkbradley committed:
removed Path import from DefaultReadWriteTest

jkbradley commented:
Rebased to pull PipelineSuite from master.
I removed the failing part of the PipelineSuite.  It was not really needed since other tests already check that save/load works and that stages are saved to appropriate directories.

SparkQA commented:
**[Test build #46110 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/46110/consoleFull)** for PR 9749 at commit [`1f61d86`](https://github.com/apache/spark/commit/1f61d8609e48314585c7171c3c86d70e10fde6cd).

mengxr commented:
Merged into master and branch-1.6. Thanks!
